File: scripts/probes/probe-ln-perform.m
//
// probe-ln-perform.m
//
// Construct an LNAction for ChatKit.OpenMessageIntent and try to execute it.
// Pure Objective-C — easier than Swift's @convention(c) bridging gymnastics.
//
// Build:
//   clang -framework AppKit -framework Foundation -ObjC probe-ln-perform.m -o probe-ln-perform
//

#import <AppKit/AppKit.h>
#import <Foundation/Foundation.h>
#import <objc/runtime.h>
#import <dlfcn.h>

static NSString *chatIdentifier(NSString *guid) {
    NSArray<NSString *> *parts = [guid componentsSeparatedByString:@";"];
    return parts.count == 3 ? parts[2] : guid;
}

int main(int argc, const char * argv[]) {
    if (argc < 3) {
        printf("usage: probe-ln-perform <chatGUID> <messageGUID>\n");
        return 2;
    }
    NSString *chatGUID = @(argv[1]);
    NSString *messageGUID = @(argv[2]);
    NSString *chatID = chatIdentifier(chatGUID);
    printf("chatGUID    = %s\n", [chatGUID UTF8String]);
    printf("messageGUID = %s\n", [messageGUID UTF8String]);
    printf("chatID      = %s\n", [chatID UTF8String]);

    void *h = dlopen("/System/Library/Frameworks/AppIntents.framework/AppIntents", RTLD_NOW);
    if (!h) { printf("FATAL: %s\n", dlerror()); return 1; }

    Class LNEntityIDClass     = NSClassFromString(@"LNEntityIdentifier");
    Class LNEntityClass       = NSClassFromString(@"LNEntity");
    Class LNValueClass        = NSClassFromString(@"LNValue");
    Class LNParameterClass    = NSClassFromString(@"LNParameter");
    Class LNActionClass       = NSClassFromString(@"LNAction");
    Class LNEntityValueType   = NSClassFromString(@"LNEntityValueType");

    if (!LNEntityIDClass || !LNEntityClass || !LNValueClass || !LNParameterClass || !LNActionClass || !LNEntityValueType) {
        printf("FATAL: missing one of the LN classes\n");
        return 1;
    }

    // 1) LNEntityIdentifier(value: messageGUID, typeName: "MessageEntity")
    id entityID = [[LNEntityIDClass alloc] performSelector:@selector(initWithValue:typeName:)
                                                withObject:messageGUID
                                                withObject:@"MessageEntity"];
    printf("entityID = %s\n", [[entityID description] UTF8String]);

    // 2) LNEntity(identifier: entityID)
    id entity = [[LNEntityClass alloc] performSelector:@selector(initWithIdentifier:) withObject:entityID];
    printf("entity   = %s\n", [[entity description] UTF8String]);

    // 3) LNEntityValueType(typeName: "MessageEntity")
    id entityType = [[LNEntityValueType alloc] performSelector:@selector(initWithTypeName:)
                                                    withObject:@"MessageEntity"];
    printf("entityType = %s\n", [[entityType description] UTF8String]);

    // 4) LNValue(value: entity, valueType: entityType)
    id lnValue = [[LNValueClass alloc] performSelector:@selector(initWithValue:valueType:)
                                            withObject:entity
                                            withObject:entityType];
    printf("lnValue    = %s\n", [[lnValue description] UTF8String]);

    // 5) LNParameter(identifier: "target", value: lnValue)
    id parameter = [[LNParameterClass alloc] performSelector:@selector(initWithIdentifier:value:)
                                                  withObject:@"target"
                                                  withObject:lnValue];
    printf("parameter  = %s\n", [[parameter description] UTF8String]);

    // 6) LNAction(identifier:"OpenMessageIntent", mangledTypeName:"7ChatKit17OpenMessageIntentV", openAppWhenRun: YES, parameters: [parameter])
    // 4-arg init — performSelector only supports 2. Use NSInvocation.
    SEL actionSel = @selector(initWithIdentifier:mangledTypeName:openAppWhenRun:parameters:);
    NSMethodSignature *sig = [LNActionClass instanceMethodSignatureForSelector:actionSel];
    if (!sig) { printf("FATAL: no signature for LNAction init\n"); return 1; }
    NSInvocation *inv = [NSInvocation invocationWithMethodSignature:sig];
    inv.target = [LNActionClass alloc];
    inv.selector = actionSel;
    NSString *intentID = @"OpenMessageIntent";
    NSString *mangled = @"7ChatKit17OpenMessageIntentV";
    BOOL openWhenRun = YES;
    NSArray *params = @[parameter];
    [inv setArgument:&intentID atIndex:2];
    [inv setArgument:&mangled atIndex:3];
    [inv setArgument:&openWhenRun atIndex:4];
    [inv setArgument:&params atIndex:5];
    [inv invoke];

    void *retainedAction = NULL;
    [inv getReturnValue:&retainedAction];
    id action = (__bridge id)retainedAction;
    printf("action     = %s\n", [[action description] UTF8String]);

    if (!action) {
        printf("FATAL: action is nil\n");
        return 1;
    }

    printf("\nLNAction properties:\n");
    printf("  identifier:      %s\n", [[action valueForKey:@"identifier"] UTF8String] ?: "(null)");
    printf("  mangledTypeName: %s\n", [[action valueForKey:@"mangledTypeName"] UTF8String] ?: "(null)");
    printf("  openAppWhenRun:  %s\n", [[[action valueForKey:@"openAppWhenRun"] stringValue] UTF8String] ?: "(null)");
    id url = [action valueForKey:@"url"];
    printf("  url:             %s\n", url ? [[url description] UTF8String] : "(null)");

    // Try opening the URL via NSWorkspace
    if ([url isKindOfClass:[NSURL class]]) {
        printf("\nOpening URL...\n");
        BOOL opened = [[NSWorkspace sharedWorkspace] openURL:(NSURL *)url];
        printf("Opened: %d\n", opened);
    } else {
        printf("\naction.url is not a URL, no open attempt.\n");
    }

    // --- LNAppContext.performAction ---
    // Try to invoke the action via LNAppContext.
    Class LNAppContextClass = NSClassFromString(@"LNAppContext");
    if (LNAppContextClass) {
        id ctx = [[LNAppContextClass alloc] init];
        printf("\nLNAppContext: %s\n", [[ctx description] UTF8String]);

        // performAction:options:reportingProgress:delegate:auditToken:completionHandler:
        SEL perfSel = @selector(performAction:options:reportingProgress:delegate:auditToken:completionHandler:);
        if ([ctx respondsToSelector:perfSel]) {
            printf("LNAppContext.performAction is callable\n");
            // We'd need audit_token_t for Messages.app PID. Try without (zeros).
            audit_token_t token; memset(&token, 0, sizeof(token));

            // Find Messages.app pid for audit token
            pid_t pid = 0;
            for (NSRunningApplication *app in [[NSWorkspace sharedWorkspace] runningApplications]) {
                if ([app.bundleIdentifier isEqualToString:@"com.apple.MobileSMS"]) {
                    pid = app.processIdentifier;
                    break;
                }
            }
            printf("Messages.app pid = %d\n", pid);

            dispatch_semaphore_t sem = dispatch_semaphore_create(0);
            void (^completion)(NSError *) = ^(NSError *err) {
                printf("performAction completion: err=%s\n", err ? [[err description] UTF8String] : "(no error)");
                dispatch_semaphore_signal(sem);
            };

            // performAction:options:reportingProgress:delegate:auditToken:completionHandler:
            NSMethodSignature *psig = [ctx methodSignatureForSelector:perfSel];
            NSInvocation *pinv = [NSInvocation invocationWithMethodSignature:psig];
            pinv.target = ctx;
            pinv.selector = perfSel;
            id options = nil;
            id progress = nil;
            id delegate = nil;
            [pinv setArgument:&action atIndex:2];
            [pinv setArgument:&options atIndex:3];
            [pinv setArgument:&progress atIndex:4];
            [pinv setArgument:&delegate atIndex:5];
            [pinv setArgument:&token atIndex:6];
            [pinv setArgument:&completion atIndex:7];
            [pinv invoke];
            printf("performAction invoked, waiting for completion...\n");
            dispatch_semaphore_wait(sem, dispatch_time(DISPATCH_TIME_NOW, 3 * NSEC_PER_SEC));
            printf("Done waiting.\n");
        }
    }

    return 0;
}

File: scripts/probes/probe-ln-perform.swift
#!/usr/bin/env swift
//
// probe-ln-perform.swift
//
// Use LinkActions (LNAction) — the Objective-C bridge that AppIntents uses
// under the hood — to construct and execute ChatKit's hidden OpenMessageIntent.
//
// Found via dyld_info on /System/Library/Frameworks/AppIntents.framework/AppIntents:
//
//   LNAction.initWithIdentifier:mangledTypeName:openAppWhenRun:parameters:
//   LNParameter.initWithIdentifier:value:
//   LNValue.initWithValue:valueType:
//   LNEntity.initWithIdentifier:properties:
//   LNEntityIdentifier.initWithTypeIdentifier:instanceIdentifier:
//   LNEntityValueType.initWithTypeName:
//   LNActionExecutor.initWithAction:connection:options:
//   LNActionExecutor.perform
//   LNConnectionAction.sendToPid:bundleIdentifier:
//
// We build an LNAction("OpenMessageIntent", "7ChatKit17OpenMessageIntentV",
// openAppWhenRun: YES, parameters: [target: MessageEntity(GUID)]) and execute it.
//

import AppKit
import Foundation
import ObjectiveC

// Unbuffered stdout for easier debugging
setbuf(stdout, nil)

// Explicitly load AppIntents — the LN* classes live here.
print("Attempting dlopen of AppIntents...")
let appIntentsHandle = dlopen("/System/Library/Frameworks/AppIntents.framework/AppIntents", RTLD_NOW)
print("AppIntents handle: \(String(describing: appIntentsHandle))")
if appIntentsHandle == nil {
    if let err = dlerror() {
        print("dlerror: \(String(cString: err))")
    }
    exit(1)
}
print("dlopen OK")

if CommandLine.arguments.count < 3 {
    print("usage: probe-ln-perform.swift <chatGUID> <messageGUID>")
    exit(2)
}
let chatGUID = CommandLine.arguments[1]
let messageGUID = CommandLine.arguments[2]
print("chatGUID    = \(chatGUID)")
print("messageGUID = \(messageGUID)")

func chatIdentifier(_ guid: String) -> String {
    let parts = guid.split(separator: ";", omittingEmptySubsequences: false)
    return parts.count == 3 ? String(parts[2]) : guid
}
let chatID = chatIdentifier(chatGUID)
print("chatID      = \(chatID)")
print()

// Use objc_msgSend directly for everything — Swift's `perform` can lose retain
// counts and class_createInstance doesn't invoke +alloc which some classes need.

// Get objc_msgSend function pointer
let msgSend = dlsym(UnsafeMutableRawPointer(bitPattern: -2)!, "objc_msgSend")!

// Various typealias-cast bindings for different signatures
typealias MsgSend0 = @convention(c) (AnyObject, Selector) -> Unmanaged<AnyObject>?
typealias MsgSend1 = @convention(c) (AnyObject, Selector, AnyObject) -> Unmanaged<AnyObject>?
typealias MsgSend2 = @convention(c) (AnyObject, Selector, AnyObject, AnyObject) -> Unmanaged<AnyObject>?
typealias MsgSend4 = @convention(c) (AnyObject, Selector, AnyObject, AnyObject, Bool, AnyObject) -> Unmanaged<AnyObject>?

let msg0 = unsafeBitCast(msgSend, to: MsgSend0.self)
let msg1 = unsafeBitCast(msgSend, to: MsgSend1.self)
let msg2 = unsafeBitCast(msgSend, to: MsgSend2.self)
let msg4 = unsafeBitCast(msgSend, to: MsgSend4.self)

let SEL_alloc = NSSelectorFromString("alloc")

@inline(never)
func alloc(_ cls: AnyClass) -> AnyObject? {
    return msg0(cls, SEL_alloc)?.takeRetainedValue()
}

@inline(never)
func callInit2(_ cls: AnyClass, sel: Selector, arg1: AnyObject, arg2: AnyObject) -> AnyObject? {
    fputs("    callInit2: sel=\(NSStringFromSelector(sel))\n", stderr); fflush(stderr)
    guard let obj = alloc(cls) else { return nil }
    let result = msg2(obj, sel, arg1, arg2)
    let v = result?.takeRetainedValue()
    fputs("    callInit2 result: \(v == nil ? "nil" : "non-nil")\n", stderr); fflush(stderr)
    return v
}

@inline(never)
func callInit1(_ cls: AnyClass, sel: Selector, arg1: AnyObject) -> AnyObject? {
    fputs("    callInit1: sel=\(NSStringFromSelector(sel))\n", stderr); fflush(stderr)
    guard let obj = alloc(cls) else { return nil }
    let result = msg1(obj, sel, arg1)
    let v = result?.takeRetainedValue()
    fputs("    callInit1 result: \(v == nil ? "nil" : "non-nil")\n", stderr); fflush(stderr)
    return v
}

@inline(never)
func callInit4(_ cls: AnyClass, sel: Selector, arg1: AnyObject, arg2: AnyObject, arg3: Bool, arg4: AnyObject) -> AnyObject? {
    fputs("    callInit4: sel=\(NSStringFromSelector(sel))\n", stderr); fflush(stderr)
    guard let obj = alloc(cls) else { return nil }
    let result = msg4(obj, sel, arg1, arg2, arg3, arg4)
    let v = result?.takeRetainedValue()
    fputs("    callInit4 result: \(v == nil ? "nil" : "non-nil")\n", stderr); fflush(stderr)
    return v
}

// MARK: - Classes

guard let LNActionClass     = NSClassFromString("LNAction"),
      let LNParameterClass  = NSClassFromString("LNParameter"),
      let LNValueClass      = NSClassFromString("LNValue"),
      let LNEntityClass     = NSClassFromString("LNEntity"),
      let LNEntityIDClass   = NSClassFromString("LNEntityIdentifier"),
      let LNEntityValueType = NSClassFromString("LNEntityValueType")
else {
    print("ERROR: Missing one or more LN* classes — AppIntents may not be loaded.")
    exit(1)
}

print("LNAction:           \(LNActionClass)")
print("LNParameter:        \(LNParameterClass)")
print("LNValue:            \(LNValueClass)")
print("LNEntity:           \(LNEntityClass)")
print("LNEntityIdentifier: \(LNEntityIDClass)")
print("LNEntityValueType:  \(LNEntityValueType)")

// MARK: - Build the MessageEntity

print("\nBuilding MessageEntity...")

// LNEntityValueType.initWithTypeName:
let entityType = callInit1(LNEntityValueType, sel: NSSelectorFromString("initWithTypeName:"), arg1: "MessageEntity" as NSString)!
fputs("  entityType created\n", stderr); fflush(stderr)

// LNEntityIdentifier — try options
fputs("\n  Trying entity identifier inits...\n", stderr); fflush(stderr)
let entityID = callInit2(LNEntityIDClass, sel: NSSelectorFromString("initWithValue:typeName:"), arg1: messageGUID as NSString, arg2: "MessageEntity" as NSString)!
fputs("  entityID created\n", stderr); fflush(stderr)

// LNEntity initWithIdentifier:
let entity = callInit1(LNEntityClass, sel: NSSelectorFromString("initWithIdentifier:"), arg1: entityID)!
fputs("  entity created\n", stderr); fflush(stderr)

// LNValue.initWithValue:valueType:
let lnValue = callInit2(LNValueClass, sel: NSSelectorFromString("initWithValue:valueType:"), arg1: entity, arg2: entityType)!
fputs("  lnValue created\n", stderr); fflush(stderr)

// LNParameter.initWithIdentifier:value:
let parameter = callInit2(LNParameterClass, sel: NSSelectorFromString("initWithIdentifier:value:"), arg1: "target" as NSString, arg2: lnValue)!
fputs("  parameter created\n", stderr); fflush(stderr)

// LNAction.initWithIdentifier:mangledTypeName:openAppWhenRun:parameters:
let action = callInit4(LNActionClass,
                       sel: NSSelectorFromString("initWithIdentifier:mangledTypeName:openAppWhenRun:parameters:"),
                       arg1: "OpenMessageIntent" as NSString,
                       arg2: "7ChatKit17OpenMessageIntentV" as NSString,
                       arg3: true,
                       arg4: [parameter] as NSArray)!
fputs("  action created\n", stderr); fflush(stderr)

// Use valueForKey to extract action properties — sometimes safer than description
if let actionObj = action as? NSObject {
    if let id = actionObj.value(forKey: "identifier") as? String {
        fputs("  action.identifier = \(id)\n", stderr); fflush(stderr)
    }
    if let mtn = actionObj.value(forKey: "mangledTypeName") as? String {
        fputs("  action.mangledTypeName = \(mtn)\n", stderr); fflush(stderr)
    }
    if let openApp = actionObj.value(forKey: "openAppWhenRun") as? NSNumber {
        fputs("  action.openAppWhenRun = \(openApp.boolValue)\n", stderr); fflush(stderr)
    }
    if let url = actionObj.value(forKey: "url") as? URL {
        fputs("  action.url = \(url.absoluteString)\n", stderr); fflush(stderr)

        // Try opening the URL!
        let opened = NSWorkspace.shared.open(url)
        fputs("  Opened URL: \(opened)\n", stderr); fflush(stderr)
    } else {
        fputs("  action.url is nil\n", stderr); fflush(stderr)
    }
}

// MARK: - Approach 2: LNAppContext.performAction
print("\n--- Attempt 2: LNAppContext.performAction:options:reportingProgress:delegate:auditToken:completionHandler: ---")

if let LNAppContextClass = NSClassFromString("LNAppContext") {
    // performAction expects auditToken — we provide our own (need to get Messages.app's)
    // First, get an LNAppContext instance.
    let appCtxAlloc = (LNAppContextClass as AnyClass).alloc()
    // -init is provided
    typealias InitT = @convention(c) (AnyObject, Selector) -> Unmanaged<AnyObject>?
    let initSel = NSSelectorFromString("init")
    let initImp = method_getImplementation(class_getInstanceMethod(LNAppContextClass, initSel)!)
    let initF = unsafeBitCast(initImp, to: InitT.self)
    let appCtx = initF(appCtxAlloc as AnyObject, initSel)?.takeRetainedValue()
    print("LNAppContext instance: \(String(describing: appCtx))")
}

print("\nDone — observe Messages.app behavior. Sleep 2s to let any async dispatch complete.")
Thread.sleep(forTimeInterval: 2.0)
